### PR TITLE
Separate rl nonrl

### PR DESCRIPTION
**Pull request by ysqyang**
# Description

1) Added "trainable_policies" parameter to EnvSampler and scenario attributes;
2) Refined env sampler logic s.t. only experiences generated by trainble policies are collected during sampling;
3) Used the trainable policies attribute to filter non-trainable policies when creating the TrainingManager;
4) SC example refinement 

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: examples/supply_chain/rl/env_sampler.py [hunk 1]
```diff
@@ -259,30 +264,23 @@ def _translate_to_env_action(self, action_dict: Dict[Any, np.ndarray], event: ob
...
-                    )
-
-                    self._consumer_orders[product_unit_id] = action_number
-                    self._orders_from_downstreams[
-                        self._storage_info["facility_levels"][source_id][product_id]["skuproduct"].id
-                    ] = action_number
-
+                    if action_number:
+                        sku = self._units_mapping[unit_id][3]
+                        reward_discount = 1
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:269]: Seems like it has not been used?

File: maro/rl/rollout/env_sampler.py [hunk 2]
```diff
@@ -192,9 +192,11 @@ class AbsEnvSampler(object, metaclass=ABCMeta):
 
     Args:
         get_env (Callable[[], Env]): Function used to create the rollout environment.
-        policy_creator (Dict[str, Callable[[str], AbsPolicy]]): Dict of functions that used to get policies, specified
-            by policy names.
-        agent2policy (Dict[Any, str]): Mapping of agent name and policy name.
+        policy_creator (Dict[str, Callable[[str], AbsPolicy]]): Dict of functions to create policies by name.
+        agent2policy (Dict[Any, str]): Mapping of agent name to policy name.
+        trainable_policies (List[str], default=None): List of trainable policy names. Experiences generated using the
```
Comment (Jinyu-W) [inline maro/rl/rollout/env_sampler.py:197]: Can we add a property to `AbsPolicy`, and let the property indicate whether it is trainable or not?

**Review by Jinyu-W — COMMENTED** [review]
You can add detailed PR descriptions to illustrate the main changes.

File: maro/rl/rollout/env_sampler.py (hunk 2 above)
Comment (ysqyang) [reply maro/rl/rollout/env_sampler.py:197]: It is not a property. Even an RLPolicy can be fixed (e.g., when using some pre-trained RL policy). We let the users define which policies they want to train.

File: examples/supply_chain/rl/env_sampler.py (hunk 1 above)
Comment (ysqyang) [reply examples/supply_chain/rl/env_sampler.py]: Removed

File: examples/cim/rl/env_sampler.py [hunk 3]
```diff
@@ -84,10 +84,15 @@ def _post_eval_step(self, cache_element: CacheElement, reward: Dict[Any, float])
...
+def env_sampler_creator(
+    policy_creator: Dict[str, Callable[[str], RLPolicy]],
+    agent2policy: Dict[str, str],
+    trainable_policies: List[str],
+) -> CIMEnvSampler:
     return CIMEnvSampler(
         get_env=lambda: Env(**env_conf),
         policy_creator=policy_creator,
         agent2policy=agent2policy,
+        trainable_policies=trainable_policies,
```
Comment (lihuoran) [inline examples/cim/rl/env_sampler.py:96]: Add a default value (None) for `trainable_policies`?

File: examples/cim/rl/env_sampler.py [hunk 4]
```diff
@@ -84,10 +84,15 @@ def _post_eval_step(self, cache_element: CacheElement, reward: Dict[Any, float])
 agent2policy = {agent: f"{algorithm}_{agent}.policy" for agent in Env(**env_conf).agent_idx_list}
```
Comment (lihuoran) [inline examples/cim/rl/env_sampler.py:84]: If we need to pass `agent2policy` as a parameter when calling `env_sampler_creator`, I suggest we move `agent2policy` to another proper place.

File: maro/rl/training/training_manager.py [hunk 5]
```diff
@@ -51,9 +51,7 @@ def __init__(
 
         self._agent2trainer: Dict[Any, str] = {}
         for agent_name, policy_name in self._agent2policy.items():
-            trainer_name = extract_trainer_name(policy_name)
-            if trainer_name in self._trainer_dict:
-                self._agent2trainer[agent_name] = trainer_name
+            self._agent2trainer[agent_name] = extract_trainer_name(policy_name)
```
Comment (lihuoran) [inline maro/rl/training/training_manager.py:54]: I think maybe we can add a check here to ensure every policy has exactly one corresponding trainer?

File: maro/rl/workflows/scenario.py [hunk 6]
```diff
@@ -18,8 +18,9 @@ def __init__(self, path: str) -> None:
         sys.path.insert(0, os.path.dirname(path))
         self._module = importlib.import_module(os.path.basename(path))
 
-    def get_env_sampler(self, policy_creator: Dict[str, Callable[[str], RLPolicy]]) -> AbsEnvSampler:
-        return getattr(self._module, "env_sampler_creator")(policy_creator)
+    @property
+    def env_sampler_creator(self) -> AbsEnvSampler:
```
Comment (lihuoran) [inline maro/rl/workflows/scenario.py:22]: `env_sampler_creator` returns `AbsEnvSampler` or `Callable`?

File: examples/cim/rl/env_sampler.py (hunk 4 above)
Comment (ysqyang) [reply examples/cim/rl/env_sampler.py]: Restored to the original signature

File: maro/rl/workflows/scenario.py (hunk 6 above)
Comment (ysqyang) [reply maro/rl/workflows/scenario.py]: Fixed

File: maro/rl/training/training_manager.py (hunk 5 above)
Comment (ysqyang) [reply maro/rl/training/training_manager.py]: Added.

**Pull request merged**